"Meta viewport allows for zoom" (b4f0c3): Explicit meaning of 'has' (#1994)
auto-wcag/auto-wcag: _rules/meta-viewport-b4f0c3.md

This rule applies to each `content` attribute on a `meta` element with a `name` [attribute value][] of `viewport` for which at least one of the following is true:

- the `content` attribute specified value includes the text `user-scalable`; or
- the `content` attribute specified value includes the text `maximum-scale`.

**Note:** The Applicability refers to string inclusion of the indicated strings in the value of the `content` attribute, as written.

## Expectation 1

For each test target, the [attribute value][] does not have a `user-scalable` property with a value of `no`.

> The [attribute value][] of the `content` attribute is a list of key-value pairs, obtained by parsing the specified value, then applying the [translations into `@viewport` descriptors][descriptor translation], notably [the translation for `user-scalable`](https://www.w3.org/TR/css-device-adapt-1/#user-scalable).

Reviewer: I'm finding this has the potential to confuse people. By referencing the transplantation to @viewport blocks, it becomes harder to understand what really is checked. A specified attribute of `no` originates an attribute value of `no` which gets translated into `fixed` for the `user-zoom` descriptor. 

The descriptor will never have a value of `no`, and is not what the expectation looks at. Therefore, I'm not sure this sentence adds value to understanding the expectation. 

Same applies to Expectation 2.

Author: Good points.
* Expectation 2, and the way it is written ("does not have a value less than 2") is pretty much the reason why we need to refer to these translations, since we want, e.g. `maximum-scale=-1` to pass (Passed Example 3). I am not sure we can do without referring to the translation one way or another.
* Expectation 1 is more interesting since it seems to be yet another meaning of "have" 🙈
* Notably, there are no example to show whether we wanted to accept `user-scalable=0`. It is "specified" as `0`, but "computed" as `fixed` and therefore passes the rule (it does not have a value of `no` in any decent sense of the term because "translated into the same value as `no` is translated" is definitely not a decent sense for "have"…) but has the exact same effect as `user-scalable=no` which fails it 💥 

I think we messed up in the latest update and Expectation 1 should say "does not have a `user-scalable` property with a value of **`fixed`**." Plus keep the note about "have" meaning "after translation" (which would match Expectation 2)

Reviewer: But from what I understand (please, correct me if I'm wrong), it's not the `user-scalable` property that gets the value fixed. It's the `user-zoom` descriptor that gets that value. So, I guess we have two choices:

1. We mention the `user-zoom` descriptor in the expectation
2. We update it to say that `user-scalable` property must not have a value of `no`, or in the range of -1 to 1, or an unknown value.

Author: Ah, yes, I see your point.
I think we should avoid rephrasing the translation, and that was a bit the goal of #1822 (no need to duplicate that and if CSS decides to change it, the rule should probably follow suit).

What about adding something like "…and using the value obtained for `user-zoom`" at the end of the sentence?

Reviewer: That could work

> The [attribute value][] of the `content` attribute is a list of key-value pairs, obtained by parsing the specified value, then applying the [translations into `@viewport` descriptors][descriptor translation], notably [the translation for `user-scalable`](https://www.w3.org/TR/css-device-adapt-1/#user-scalable), and using the value obtained for `user-zoom`.

Reviewer: Trying to make the algorithm clearer. Also adding the "For the purpose of this rule" at the beginning so that it is clear that the attribute value is not computed using our standard definition. 

```suggestion
For the purpose of this rule, the [attribute value][] of the `content` attribute is a list of key-value pairs, obtained by:

1. parsing the specified value for each key-value pair;
2. applying the [translations into `@viewport` descriptors][descriptor translation];
3. using the value of the corresponding `@viewport` descriptor, notably using the value obtained for the `user-zoom` descriptor from [the translation for `user-scalable`](https://www.w3.org/TR/css-device-adapt-1/#user-scalable).
```

## Expectation 2

For each test target, the [attribute value][] does not have a `maximum-scale` property with a value less than 2.

> The [attribute value][] of the `content` attribute is a list of key-value pairs, obtained by parsing the specified value, then applying the [translations into `@viewport` descriptors][descriptor translation], notably [the translation for `maximum-scale`](https://www.w3.org/TR/css-device-adapt-1/#min-scale-max-scale) (including dropping the key-value pair completely for invalid values).

Reviewer: Same issue here. What we want is the `max-zoom` description to not have a value less than 2. So, we can either say that in the expectation, or say that the `maximum-scale` property, must not have a numeric value under 2, or a literal value that is different from `device-width` and `device-height`

Reviewer: Still, if we have an example like 

```
<meta name="viewport" content="maximum-scale=device-width">
```

how do I compare `device-width` with `2`?

Author: After translating it and using the value for `max-zoom`, `device-width` results in a `max-zoom` of 10, which is greater than 2, so this is OK 🤔

Reviewer: I think only now it clicked in my head. I was using the definition of `attribute value` to compute the value. But the second paragraph adds to the definition. Should we, instead, update the definition of `attribute value` to compute the value in this instance that refers to the value of a CSS property instead of the value of the HTML attribute?

Author: I'm not sure because this is currently the only rule doing that, so I'd prefer to keep locality.
OTOH, we can definitely  improve the situation with some examples. I'll give it a go.

> The [attribute value][] of the `content` attribute is a list of key-value pairs, obtained by parsing the specified value, then applying the [translations into `@viewport` descriptors][descriptor translation], notably [the translation for `maximum-scale`](https://www.w3.org/TR/css-device-adapt-1/#min-scale-max-scale) (including dropping the key-value pair completely for invalid values), and using the value obtained for `max-zoom`.

Reviewer: ```suggestion
For the purpose of this rule, the [attribute value][] of the `content` attribute is a list of key-value pairs, obtained by:

 1. parsing the specified value for each key-value pair;
 2. applying the [translations into `@viewport` descriptors][descriptor translation];
 3. using the value of the corresponding `@viewport` descriptor  (including dropping the key-value pair completely for invalid values), notably using the value obtained for the `max-zoom` descriptor from [the translation for `maximum-scale`](https://www.w3.org/TR/css-device-adapt-1/#min-scale-max-scale).
```

This rule applies to each `content` attribute on a `meta` element with a `name` [attribute value][] of `viewport` for which at least one of the following is true:

- the `content` attribute specified value includes the text `user-scalable`; or
- the `content` attribute specified value includes the text `maximum-scale`.

**Note:** The Applicability refers to string inclusion of the indicated strings in the value of the `content` attribute, as written.

## Expectation 1

> For each test target, the [attribute value][] does not have a `user-scalable` property with a value of `fixed`.

Reviewer 2: Should this have been `user-zoom`?

I'm a little hesitant for us to start using the CSS property names / values here in place of the actual key/value pairs on the content attribute. It creates adds a layer of abstraction to the rule 
which makes it more difficult to understand. Since rules on how to process these props are pretty easy, I would suggest we put them straight into the rule in the same way that the autocomplete rule has a pseudo-copy of the HTML spec's description of what makes a valid autocomplete attribute value.

```suggestion
For each test target, when the [attribute value][] is parsed into key/value pairs, one of the following is true:
- there is no `user-scalable` key; or
- the `user-scalable` value is `yes`; or
- the `user-scalable` value is a number less than -1, or greater than 1.
```

We can then put something in the background about how the expectation follows the steps described in CSS. You also wouldn't have to add those translate statements to the example descriptions.

Author: Good idea. I gave it a shot 👍

## Applicability

This rule applies to each `content` attribute for which all of the following are true:

- the attribute is defined on a `meta` element with a `name` [attribute value][] of `viewport`; and

> - the [attribute value][] has either the `maximum-scale` or `user-scalable` key.

Reviewer: Doesn't this make the following attribute inapplicable?

```html
<meta name="viewport" content="maximum-scale=2, user-scalable=5" />
```

It doesn't have either of the attributes. It has both. 
And I believe this should be applicable.

For the purpose of this rule, the [attribute value][] of the `content` attribute is a list of key/value pairs.

## Expectation 1

For each test target's [attribute value][], at least one of the following is true:

> - the `user-scalable` key is not defined; or
> - the `user-scalable` key is paired with a value of `yes`, `device-width` or `device-height`; or
> - the `user-scalable` key is paired with a numeric value which is not in the range ]-1;1[.

Reviewer 2: I would prefer we simplify this a bit. I don't think this is any less ambiguous than what you had:

```suggestion
- `user-scalable` is not defined; or
- `user-scalable` is `yes`, `device-width` or `device-height`; or
- `user-scalable` is a number which is not between -1 and 1.
```

## Applicability

This rule applies to each `content` attribute for which all of the following are true:

- the attribute is defined on a `meta` element with a `name` [attribute value][] of `viewport`; and
- the [attribute value][] has at least one of the `maximum-scale` or `user-scalable` keys.

For the purpose of this rule, the [attribute value][] of the `content` attribute is a list of key/value pairs.

## Expectation 1

For each test target's [attribute value][], at least one of the following is true:

> - `user-scalable` is not defined; or

Reviewer 2: I'm okay with this as you write it, but I think it might be better if eventually we split this up. Have one rule for user-scalable, and a second for maximum-scale. This split expectation is more complicated than it needs to be I think because of this.

Author: Good point.

Related to that, and given what we semi-often do, I wonder if it would be a good idea to allow expectations like:
> For each test target [which verify some condition], …

So that test targets that do not verify the condition would automatically pass that expectation.

In that case, that would allow to write:
> For each test target that defines `user-scalable`, …

And in general, anytime we have Applicability with an alternative, that would allow somewhat simpler expectation. And result in less rules overall, I feel these splittings are often diluting rules that should stay close together.

Alternatively, this could also be done with Composite rules. Having one Atomic rule for each property, and a Composite requesting both to not fail. It is a bit redundant, but avoids splitting rules that are really working together.

Author: Tracking the idea in https://github.com/w3c/wcag-act/issues/548

#### Passed Example 2

This viewport `meta` element does not prevent user scaling because it has `user-scalable` set to `5`.

```html
<html>
	<head>
		<title>Simple page showing random text</title>
		<meta name="viewport" content="user-scalable=5" />
	</head>
	<body>
		<p>
			Lorem ipsum
		</p>
	</body>
</html>

> ```

Reviewer 3: Is there a reason that this new example was added in the middle rather than the end? I'm not sure that we're strict about this, but if there's something out there already that refers to Pass Example 2, that reference would be bad since Pass Example 2 is now Pass Example 3. I'm sure it happens anyway when we remove examples, but I don't see a reason to do that in this case.

Author: Yes, I wanted to keep the `user-scalable` tests together, and the `maximum-scale` together. I think it makes the rule more readable than if we mix them (less need to switch context).

Our test cases numbers are absolutely not stable, and we should not link to them directly. This is acceptable when referring to them in issues, … but we shouldn't, e.g., have a rule linking to an example with a `passed-example-2` id. If we need to link inside rule, we should manually had an anchor.
